src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,731 @@
+          })
+          .cloned()
+          .collect();
+        for file_path in matching_files.into_iter() {
+          // NB: remove any matched files from `cur_dir_files`, so they are successfully matched
+          // against at most once.
+          let file_node = cur_dir_files.remove(&file_path).unwrap();
+          globbed_files.insert(file_path, file_node);
+        }
+
+        let matching_directories: Vec<PathBuf> = cur_dir_directories
+          .keys()
+          .filter(|path| {
+            // NB: match just the current path component against the wildcard, but use the prefix
+            // when checking against the `exclude` patterns!
+            wildcard.matches_path(path) && !exclude.is_ignored_path(&prefix.join(path), true)
+          })
+          .cloned()
+          .collect();
+        for directory_path in matching_directories.into_iter() {
+          // NB: do *not* remove the directory for `cur_dir_directories` after it is matched
+          // successfully once!
+          let directory_node = cur_dir_directories.get(&directory_path).unwrap();
+
+          // TODO(#9967): Figure out how to consume the existing glob matching logic that works on
+          // `VFS` instances!
+          match &path_glob {
+            RestrictedPathGlob::Wildcard { wildcard } => {
+              assert_ne!(*wildcard, *DOUBLE_STAR_GLOB);
+              // This directory matched completely, without having to subset its contents
+              // whatsoever. We can avoid traversing its contents and return the node.
+              // NB: This line should be idempotent if the same directory node is added twice.
+              globbed_directories.insert(directory_path, directory_node.clone());
+            }
+            RestrictedPathGlob::DirWildcard {
+              wildcard,
+              remainder,
+            } => {
+              let mut subdir_globs: Vec<RestrictedPathGlob> = vec![];
+              if *wildcard == *DOUBLE_STAR_GLOB {
+                // NB: here we short-circuit several cases which would swallow up a directory
+                // without subsetting it.
+                match &remainder[..] {
+                  // ending in /**
+                  [] => {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  // ending in /**/* or /**/**
+                  [single_glob]
+                    if *single_glob == *SINGLE_STAR_GLOB || *single_glob == *DOUBLE_STAR_GLOB =>
+                  {
+                    globbed_directories.insert(directory_path, directory_node.clone());
+                    continue;
+                  }
+                  _ => (),
+                }
+                let with_double_star = RestrictedPathGlob::DirWildcard {
+                  wildcard: wildcard.clone(),
+                  remainder: remainder.clone(),
+                };
+                subdir_globs.push(with_double_star);
```

Comment:
This adds to `subdir_globs` unconditionally, but we should only continue to recurse if the remainder is not empty, right? If the remainder is empty, we should shortcircuit recursion.

The block below this appears to account for that: `match remainder.len()`...

Reply:
I've made this more clear in cleaning up the code -- whenever `*wildcard` is a `**` and `remainder` is non-empty, we will always continue to recurse, and we will have two branches, one for the remainder alone, and one for the `**` prepended to the remainder.

Reply:
I have added testing to check that the recursion short-circuiting is occurring!